docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,211 @@
+Embed APIv3
+===========
+
+The Embed API allows to embed content from documentation pages in other sites.
+It has been treated as an *experimental* feature without public documentation or real applications,
+but recently it started to be used widely (mainly because we created a Sphinx extension).
+
+The main goal of this document is to design a new version of the Embed API to be more user friendly,
+make it more stable over time, support documentation pages not hosted at Read the Docs,
+and remove some quirkiness that makes it hard to maintain and difficult to use.
+
+.. note::
+
+   This work is part of the `CZI grant`_ that Read the Docs received.
+
+.. _CZI grant: https://blog.readthedocs.com/czi-grant-announcement/
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Current implementation
+----------------------
+
+The current implementation of the API is partially documented in :doc:`/guides/embedding-content`.
+It has some known problems:
+
+* There are different ways of querying the API: ``?url=`` (generic) and ``?doc=`` (relies on Sphinx's specific concept)
+* Doesn't support MkDocs
+* Lookups are slow (~500 ms)
+* IDs returned aren't well formed (like empty IDs ``"headers": [{"title": "#"}]``)
+* The content is always an array of one element
+* It tries different variations of the original ID
+* It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag)
+
+
+Goals
+-----
+
+Considering the problems mentioned in the previous section,
+the inclusion of new features and the definition of a contract that works the same for all,
+this document set the following goals for the new version of this endpoint:
+
+* Support external documents hosted outside Read the Docs
+* Do not depend on Sphinx ``.fjson`` files
+* Query and parse the ``.html`` file directly (from our storage or from an external request)
+* Rewrite all links returned in the content to make them absolute
+* Always return valid HTML structure
+* Delete HTML tags from the original document if needed
+* Support ``?nwords=`` and ``?nparagraphs=`` to return chunked content
```

Comment:
I think we still were discussing this. I don't think we can support it without having to handle specific cases, not all returned html will have `p` tags or the word will end in a valid tag, we also need to skip tags from the lenght


```
<p>Cut <strong>here?</strong></p>
```

Having the client using css to "hide" long texts is more easy to handle.

Reply:
The example you share would return:

1. with `?nwords=1`: `<p>Cut</p>`
2. with `?nwords=2`: `<p>Cut <strong>here?</strong></p>`
3. with `?nparagraphs=1`: `<p>Cut <strong>here?</strong></p>`

Reply:
yeah, but the problem is calculating the words and taking the tags into account, you'll be removing tags from the original html, possible breaking some styles like with tables/definition lists. And also how would you know if there is a paragraph when you have content surrounded in other tags like lists or divs.

Reply:
We have been playing with this already in https://github.com/readthedocs/readthedocs-ext/pull/304/. _How to do it_ is an implementation detail and I'm sure there is going to be some problems we will need to solve. However, I don't think it's impossible.

I've been playing a little with BeautifulSoup already and this seems to work close enough:

```python
# nparagraphs.py
from bs4 import BeautifulSoup

soup = BeautifulSoup(open('install.html'))
nparagraphs = 3
for element in soup.find('div', attrs={'id': 'development-installation'}).findAll():
    if nparagraphs == 0:
        element.replaceWith('')

    if element.name == 'p' and nparagraphs > 0:
        nparagraphs -= 1
```

```bash
$ wget https://docs.readthedocs.io/en/stable/development/install.html
$ python nparagraph.py
```

<details>
  <summary>Click to see the output</summary>

```html
<div class="section" id="development-installation">
 <h1>
  Development Installation
  <a class="headerlink" href="#development-installation" title="Permalink to this headline">
   ¶
  </a>
 </h1>
 <p>
  These are development setup and
  <a class="hoverxref tooltip reference internal" data-doc="development/install" data-docpath="/development/install.html" data-project="docs" data-section="core-team-standards" data-version="stable" href="#core-team-standards">
   <span class="std std-ref">
    standards
   </span>
  </a>
  that are adhered to by the core development team while
developing Read the Docs and related services. If you are a contributor to Read the Docs,
it might a be a good idea to follow these guidelines as well.
 </p>
 <div class="admonition note">
  <p class="admonition-title">
   Note
  </p>
  <p>
   We do not recommend to follow this guide to deploy an instance of Read the Docs for production usage.
Take into account that this setup is only useful for developing purposes.
  </p>
 </div>
</div>
```
</details>